### Importar las librerías

In [40]:
import pandas as pd

### Cargar el archivo chipotle.csv en un dataframe y guardarlo en una variable de nombre chipo

In [41]:
chipo = pd.read_csv('./chipotle.csv')

### Ver las primeras 10 filas

In [42]:
chipo.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                                NaN     $1.69   
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25   
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25

In [43]:
chipo[chipo.item_name == "Veggie Salad Bowl"].quantity.sum()

18

### ¿Cuál es el número de muestras en el dataset? ¿Y el número de atributos? Mostrar el nombre de los atributos

In [44]:
chipo.shape
chipo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
order_id              4622 non-null int64
quantity              4622 non-null int64
item_name             4622 non-null object
choice_description    3376 non-null object
item_price            4622 non-null object
dtypes: int64(2), object(3)
memory usage: 180.6+ KB


### ¿Cuántos items se ordenaron en total? Notar que una fila puede contener más de un item en la columna *quantity*

In [45]:
# chipo.quantity.value_counts()

chipo.quantity.sum()

4972

### ¿Cuántas órdenes se hicieron en total? Notar que en una orden se pueden comprar muchos productos. Hay una orden por cada *orden_id*

In [46]:
# chipo.order_id.nunique()

chipo.order_id.max() # asumiendo que no falta ningun orden_id

# chipo.groupby('order_id').count().shape
# chipo.order_id.unique()

1834

### ¿Cuántos items diferentes se venden?

In [47]:
# chipo.item_name.nunique()
# len(chipo.item_name.unique())
chipo.item_name.unique().shape

(50,)

### Convertir el atributo *item_price* a float. Tener en cuenta que el formato es un string de la forma "$2.39 " por lo que necesitamos remover el primer y el último caracter del string. **IMPORTANTE PARA SEGUIR**

In [52]:
chipo['item_price'] = chipo['item_price'].replace( '[\$,)]','', regex=True, inplace=True)

In [53]:
chipo.item_price.head()

0     2.39 
1     3.39 
2     3.39 
3     2.39 
4    16.98 
Name: item_price, dtype: object

In [34]:
print(type(chipo['item_price']), chipo.item_price.dtype)

<class 'pandas.core.series.Series'> object


In [20]:
chipo['item_price'] = chipo['item_price'].str[1:-1].astype('float')
# chipo.item_price.dtype

¿De qué otra manera podemos referirnos a la columna *item_price* del dataframe?

In [21]:
chipo.item_price
chipo.loc[:, 'item_price']

0        2.39
1        3.39
2        3.39
3        2.39
4       16.98
5       10.98
6        1.69
7       11.75
8        9.25
9        9.25
10       4.45
11       8.75
12       8.75
13      11.25
14       4.45
15       2.39
16       8.49
17       8.49
18       2.18
19       8.75
20       4.45
21       8.99
22       3.39
23      10.98
24       3.39
25       2.39
26       8.49
27       8.99
28       1.09
29       8.49
        ...  
4592    11.75
4593    11.75
4594    11.75
4595     8.75
4596     4.45
4597     1.25
4598     1.50
4599     8.75
4600     4.45
4601     1.25
4602     9.25
4603     9.25
4604     8.75
4605     4.45
4606     1.25
4607    11.75
4608    11.25
4609     1.25
4610    11.75
4611    11.25
4612     9.25
4613     2.15
4614     1.50
4615     8.75
4616     4.45
4617    11.75
4618    11.75
4619    11.25
4620     8.75
4621     8.75
Name: item_price, Length: 4622, dtype: float64

### ¿En qué consistió la orden número 1000? ¿Cuánto gastó?

In [31]:
# aux = chipo[chipo.order_id == 1000]
# (aux.item_price * aux.quantity).sum()
display(aux)

order_id  quantity      item_name  \
2517      1000         1  Barbacoa Bowl   
2518      1000         1   Chicken Bowl   

                                     choice_description  item_price  
2517  [Roasted Chili Corn Salsa, [Rice, Black Beans,...        9.25  
2518  [Fresh Tomato Salsa, [Black Beans, Cheese, Gua...       11.25

### ¿Cuántas *Veggie Salad Bowl* se compraron?

In [44]:
# chipo[chipo.item_name == "Veggie Salad Bowl"]#.quantity.sum()
# chipo[chipo.item_name == "Veggie Salad Bowl"]
# chipo.loc[chipo.item_name == "Veggie Salad Bowl", :]

# chipo.query("item_name == 'Veggie Salad Bowl'")

### ¿Cuántas veces alguien ordenó más de una *Canned Soda*?

¿Cuántas veces en una fila hay más de una *Canned Soda* en *quantity*?

In [46]:
len(chipo[(chipo.item_name == "Canned Soda") & (chipo.quantity > 1)])#.shape[0]


# aux = chipo.groupby(['order_id', 'item_name']).quantity.sum()
# aux = aux.reset_index()
# aux[(aux.item_name == "Canned Soda") & (aux.quantity > 1)].shape[0]

20

¿Cuántas veces alguien ordenó más de una *Canned Soda* teniendo en cuenta que en una orden puede haber más de una separada?

In [53]:
# chipo.groupby('order_id').sum() # notar que order_id pasa a ser index

chipo_soda = chipo[chipo.item_name == "Canned Soda"]

chipo_soda.groupby(['order_id']).sum().quantity

soda_orders = chipo_soda.groupby(['order_id']).sum().quantity
soda_orders[soda_orders > 1].count()

24

### Agregar una columna de nombre *revenue* (ingreso) que contenga el dinero ganado por la compra de los items de la fila

In [57]:
# chipo['revenue'] = chipo['quantity'] * chipo['item_price']
chipo.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description  item_price  revenue  
0                                                NaN        2.39     2.39  
1                                       [Clementine]        3.39     3.39  
2                                            [Apple]        3.39     3.39  
3                                                NaN        2.39     2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98    33.96  
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...       10.98    10.98  
6                                                NaN        1.69     1.69  
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...       11.75    11.75  
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...        9.25     9.25  
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...        9.25     9.25

### ¿Cuál es el *revenue* (ingreso) promedio por orden?

In [60]:
# chipo.groupby(['order_id']).sum()

chipo.groupby(['order_id']).sum().mean()['revenue']

21.394231188658654

### Mostrar las 3 filas que dieron como resultado el mayor *revenue*

In [63]:
chipo.sort_values
chipo.sort_values(by="revenue", ascending=False).head(3)

order_id  quantity                     item_name  \
3598      1443        15  Chips and Fresh Tomato Salsa   
4152      1660        10                 Bottled Water   
1254       511         4               Chicken Burrito   

                                     choice_description  item_price  revenue  
3598                                                NaN       44.25   663.75  
4152                                                NaN       15.00   150.00  
1254  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       35.00   140.00

### ¿Cuál fue el item más ordenado?

In [70]:
# chipo.groupby('item_name').sum()

# chipo.groupby('item_name').sum().sort_values(['quantity'], ascending=False).head(1)
# chipo.groupby('item_name').sum().quantity.idxmax()

### ¿Qué ingresos produjo el item más ordenado?

In [74]:
mas_ordenado = chipo.groupby('item_name').sum().quantity.idxmax()

# chipo.groupby('item_name').sum()

# chipo.groupby('item_name').sum().loc[mas_ordenado].revenue

# chipo[chipo.item_name == mas_ordenado].revenue.sum()

order_id  quantity  item_price  \
item_name                                                                  
6 Pack Soft Drink                       968.925926  1.018519    6.610185   
Barbacoa Bowl                           817.757576  1.000000   10.187273   
Barbacoa Burrito                        821.076923  1.000000    9.832418   
Barbacoa Crispy Tacos                   510.272727  1.090909   10.928182   
Barbacoa Salad Bowl                     970.800000  1.000000   10.640000   
Barbacoa Soft Tacos                     749.000000  1.000000   10.018400   
Bottled Water                          1086.074074  1.302469    1.867654   
Bowl                                    236.000000  2.000000   14.800000   
Burrito                                 258.333333  1.000000    7.400000   
Canned Soda                             734.576923  1.211538    1.320577   
Canned Soft Drink                      1012.468439  1.166113    1.457641   
Carnitas Bowl                           922.676471  1.044118   10.833971   
Carnitas Burrito                        875.457627  1.016949   10.132712   
Carnitas Crispy Tacos                   838.285714  1.142857   11.137143   
Carnitas Salad                         1500.000000  1.000000    8.990000   
Carnitas Salad Bowl                    1115.500000  1.000000   11.056667   
Carnitas Soft Tacos                     941.825000  1.000000    9.398500   
Chicken Bowl                            983.369146  1.048209   10.113953   
Chicken Burrito                         899.282098  1.068716   10.082857   
Chicken Crispy Tacos                   1008.127660  1.063830   10.045319   
Chicken Salad                          1162.777778  1.000000    9.010000   
Chicken Salad Bowl                     1064.581818  1.118182   11.170455   
Chicken Soft Tacos                      855.608696  1.043478    9.635565   
Chips                                   985.800948  1.090047    2.342844   
Chips and Fresh Tomato Salsa            912.900000  1.181818    3.285091   
Chips and Guacamole                     939.371608  1.056367    4.595073   
Chips and Mild Fresh Tomato Salsa       279.000000  1.000000    3.000000   
Chips and Roasted Chili Corn Salsa      681.772727  1.045455    3.084091   
Chips and Roasted Chili-Corn Salsa      970.111111  1.000000    2.390000   
Chips and Tomatillo Green Chili Salsa   957.441860  1.046512    3.087209   
Chips and Tomatillo Red Chili Salsa     949.083333  1.041667    3.072917   
Chips and Tomatillo-Green Chili Salsa   704.064516  1.064516    2.544194   
Chips and Tomatillo-Red Chili Salsa     917.200000  1.250000    2.987500   
Crispy Tacos                            217.000000  1.000000    7.400000   
Izze                                    529.350000  1.000000    3.390000   
Nantucket Nectar                        807.925926  1.074074    3.641111   
Salad                                   575.000000  1.000000    7.400000   
Side of Chips                           839.297030  1.089109    1.840594   
Steak Bowl                              918.255924  1.047393   10.711801   
Steak Burrito                           892.491848  1.048913   10.465842   
Steak Crispy Tacos                     1023.371429  1.028571   10.209714   
Steak Salad                             590.500000  1.000000    8.915000   
Steak Salad Bowl                       1020.344828  1.068966   11.847931   
Steak Soft Tacos                        912.309091  1.018182    9.746364   
Veggie Bowl                             850.235294  1.023529   10.211647   
Veggie Burrito                          852.231579  1.021053    9.839684   
Veggie Crispy Tacos                     668.000000  1.000000    8.490000   
Veggie Salad                           1210.833333  1.000000    8.490000   
Veggie Salad Bowl                       961.277778  1.000000   10.138889   
Veggie Soft Tacos                       788.571429  1.142857   10.565714   

                                         revenue  
item_name                             

### Chipotle decide abrir su primera sucursal en Argentina. Para eso nos pide que realicemos la lista de precios de sus productos

#### Usando nuestro dataframe chipo, agrupemos por *item_name* y calculemos la media de *item_price* para cada item. Creemos un nuevo dataframe de nombre *chipo_precios* que solo contenga los *item_name* como index y una columna con los *item_price*.

In [77]:
chipo_precios = chipo.groupby('item_name').mean()

chipo_precios = chipo_precios.drop(columns=['order_id', 'quantity', 'revenue'])

chipo_precios = chipo_precios.round({'item_price': 2})

# chipo_precios

#### Chipotle decide reemplazar su producto más ordenado (Chicken Bowl) por algo más local (Chori Bowl). Remover de *chipo_precios* la fila correspondiente y agregar usando concat o append el flamante Chori Bowl que se venderá al módico precio de $5 (dólares). Ordenar *chipo_precios* por index (item_name)

In [80]:
# chori = pd.DataFrame([5.], index=['Chori Bowl'], columns=['item_price'])
# chipo_precios = pd.concat([chipo_precios, chori])
# chipo_precios = chipo_precios.drop('Chicken Bowl', axis=0)
# chipo_precios = chipo_precios.sort_index()
# # chipo_precios

#### Agregar una columna a nuestro dataframe de nombre *item_price_arg* que obtenga el valor en pesos argentinos de cada producto usando la columna *item_price* y el valor del dólar hoy respecto al peso argentino

In [ ]:
# crear la serie y concat
chipo_precios['item_price_arg'] = chipo_precios['item_price'] * 38
chipo_precios

#### Chipotle ve el valor del dólar en Enero 2018 y decide instalarse en Uruguay. Renombrar la columna anterior a *item_price_uru* usando rename y recalcular sus valores usando el valor del peso uruguayo

In [ ]:
chipo_precios = chipo_precios.rename(columns={'item_price_arg': 'item_price_uru'})
chipo_precios['item_price_uru'] = chipo_precios['item_price'] * 33
chipo_precios